### PR TITLE
[SPARK-37155][PYTHON] Inline type hints for python/pyspark/statcounter.py

### PR DESCRIPTION
**Pull request by dchvn**
### What changes were proposed in this pull request?
Inline type hints for python/pyspark/statcounter.py

### Why are the changes needed?
We can take advantage of static type checking within the functions by inlining the type hints.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests

### CODE REVIEW

**Review by HyukjinKwon — COMMENTED** [review]
looks fine

**Comment by dchvn** [comment]
Ping @HyukjinKwon @zero323 Thanks !

File: python/pyspark/statcounter.py
Comment (HyukjinKwon) [inline python/pyspark/statcounter.py:68]: this looks weird. when other is self, shouldn't `other` be `StatCounter`?

File: python/pyspark/statcounter.py
Comment (HyukjinKwon) [inline python/pyspark/statcounter.py:26]: can we avoid this?

File: python/pyspark/statcounter.py
Comment (zero323) [inline python/pyspark/statcounter.py:140]: Nit. Might be cleaner to have `Dict[str, float]`.

**Comment by zero323** [comment]
In general, this should be rewritten with `Protocols` or equivalent at some point, but it is not priority and it is not in the scope of this PR.

File: python/pyspark/statcounter.py
Comment (dchvn) [reply python/pyspark/statcounter.py:140]: @zero323 I think we should keep ```Dict[str, Union[float, int]]```  because of ```'count': self.count()``` is ```int``` in the return. WDYT?

Comment (zero323) [reply python/pyspark/statcounter.py:136]: There are contexts, where distinguishing between these two is important (for us, it is primarily for arguments which are passed to JVM), however every integer is a real number (see [Numerical tower](https://en.wikipedia.org/wiki/Numerical_tower) and, albeit not that useful in practice,  [PEP 3141](https://www.python.org/dev/peps/pep-3141/)).

**Comment by dchvn** [comment]
This PR is ready to review again! Thanks, @zero323 @HyukjinKwon

File: python/pyspark/statcounter.py
Comment (dchvn) [reply python/pyspark/statcounter.py]: updated. Thanks!

File: python/pyspark/statcounter.py
Comment (dchvn) [reply python/pyspark/statcounter.py]: updated. Thanks!

File: python/pyspark/statcounter.py
Comment (dchvn) [reply python/pyspark/statcounter.py]: updated, https://github.com/apache/spark/pull/34435#discussion_r743501203. Thanks!

File: python/pyspark/statcounter.py
Comment (zero323) [reply python/pyspark/statcounter.py:29]: I don't have strong option on choosing specific option, but:

- If we're going to use typed definitions before `try` / `catch` block, we can probably do better than `Any`.
- If we're going to use ignores, these should use specific codes (`[assignment]` if I am not mistaken).

Comment (dchvn) [reply python/pyspark/statcounter.py]: Updated. Thanks!

Comment (dchvn) [reply python/pyspark/statcounter.py]: @zero323 , can you review it again? Many thanks 😄

**Comment by dchvn** [comment]
cc @zero323 , @HyukjinKwon , @ueshin !

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
im gonna leave it to @zero323 since he has been reviewing it.

File: python/pyspark/statcounter.py
Comment (ueshin) [inline python/pyspark/statcounter.py:64]: How can we cast `cast(float, copy.deepcopy(other))` here?
I guess this is a long-standing bug and should call `self.mergeStats` instead of `self.merge`? cc @HyukjinKwon

**Comment by zero323** [comment]
LGTM, assuming we don't want address https://github.com/apache/spark/pull/34435#discussion_r750668911 in this PR

File: python/pyspark/statcounter.py
Comment (zero323) [inline python/pyspark/statcounter.py:62]: Let's `type: ignore` handle this as a separate ticket. This has been here forever and was never reported, so there is good indication it is a piece of dead code, and we should cover this scenario with unit tests anyway.

Comment (dchvn) [reply python/pyspark/statcounter.py:62]: Should I just add # type: ignore like: `self.merge(copy.deepcopy(other))  # type: ignore` and create a new Jira ticket with UT for this issue? @zero323 , @ueshin , @HyukjinKwon 
Many thanks!

Comment (HyukjinKwon) [reply python/pyspark/statcounter.py:62]: Yeah, let's handle it in a separate PR.

**Review by zero323 — APPROVED** [review]

**Review by zero323 — APPROVED** [review]

**Comment by zero323** [comment]
Merged to master.

**Comment by dchvn** [comment]
Thanks all !

**Comment by ueshin** [comment]
I submitted a PR to fix the issue https://github.com/apache/spark/pull/34435#discussion_r750668911 at #34653

**Pull request closed without merging**
